src/dowel/csv_output.py

```diff
@@ -2,59 +2,66 @@
-        super().__init__(file_name)
+    def __init__(self, file_name, keys_accepted=r'^\S+$'):
+        super().__init__(file_name, keys_accepted=keys_accepted)
         self._writer = None
         self._fieldnames = None
         self._warned_once = set()
         self._disable_warnings = False
+        self.tabular = Tabular()
 
     @property
     def types_accepted(self):
-        """Accept TabularInput objects only."""
-        return (TabularInput, )
-
-    def record(self, data, prefix=''):
-        """Log tabular data to CSV."""
-        if isinstance(data, TabularInput):
-            to_csv = data.as_primitive_dict
-
-            if not to_csv.keys() and not self._writer:
-                return
-
-            if not self._writer:
-                self._fieldnames = set(to_csv.keys())
-                self._writer = csv.DictWriter(
-                    self._log_file,
-                    fieldnames=self._fieldnames,
-                    extrasaction='ignore')
-                self._writer.writeheader()
-
-            if to_csv.keys() != self._fieldnames:
-                self._warn('Inconsistent TabularInput keys detected. '
-                           'CsvOutput keys: {}. '
-                           'TabularInput keys: {}. '
-                           'Did you change key sets after your first '
-                           'logger.log(TabularInput)?'.format(
-                               set(self._fieldnames), set(to_csv.keys())))
-
-            self._writer.writerow(to_csv)
-
-            for k in to_csv.keys():
-                data.mark(k)
-        else:
-            raise ValueError('Unacceptable type.')
+        """Accept str and scalar objects."""
+        return (str, ) + np.ScalarType
+
+    def record(self, key, value, prefix=''):
+        """Log data to a csv file."""
+        self.tabular.record(key, value)
+
+    def dump(self, step=None):
+        """Flush data to log file."""
+        if self.tabular.empty:
+            return
+
+        to_csv = self.tabular.as_primitive_dict
+
+        if not to_csv.keys() and not self._writer:
+            return
+
+        if not self._writer:
+            self._fieldnames = set(to_csv.keys())
+            self._writer = csv.DictWriter(self._log_file,
+                                          fieldnames=self._fieldnames,
+                                          extrasaction='ignore')
+            self._writer.writeheader()
+
+        if to_csv.keys() != self._fieldnames:
+            self._warn('Inconsistent Tabular keys detected. '
+                       'CsvOutput keys: {}. '
+                       'Tabular keys: {}. '
+                       'Did you change key sets after your first '
+                       'logger.log(Tabular)?'.format(set(self._fieldnames),
+                                                     set(to_csv.keys())))
+
+        self._writer.writerow(to_csv)
+
+        self._log_file.flush()
+        self.tabular.clear()
```

Comment:
i don't think we should clear the table between calls to dump, because it allows us to provide a value even if someone doesn't update it. basically, if KV pairs are not all updated at the same rate it's okay, and we don't need to output an error.